ALE does not show errors for cppcheck

<!--
  This is the template for reporting ALE bugs. Make sure you try updating ALE
  to a more recent version before reporting a bug. Look through existing bug
  reports for similar issues before reporting a new one. Don't leave comments
  about new bugs in the comment section for old issues.

  Make sure to try disabling other plugins and trying to repeat your bug before
  reporting it in ALE. Some times problems can arise when two plugins are used
  together, but often your issues might be problems with other plugins.
-->

## Information

**VIM version**
```
VIM - Vi IMproved 8.1 (2018 May 18, compiled May  2 2019 12:07:21)
macOS version
Included patches: 1-1189
Compiled by rasmus@Falcon-9.local
Huge version without GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     -tag_old_static
+arabic            -farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    -tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
-balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
-browse            +fork()            -mzscheme          +textobjects
++builtin_terms    -gettext           +netbeans_intg     +textprop
+byte_offset       -hangul_input      +num64             +timers
+channel           +iconv             +packages          +title
+cindent           +insert_expand     +path_extra        -toolbar
-clientserver      +job               -perl              +user_commands
+clipboard         +jumplist          +persistent_undo   +vartabs
+cmdline_compl     +keymap            +postscript        +vertsplit
+cmdline_hist      +lambda            +printer           +virtualedit
+cmdline_info      +langmap           +profile           +visual
+comments          +libcall           -python            +visualextra
+conceal           +linebreak         +python3           +viminfo
+cryptv            +lispindent        +quickfix          +vreplace
-cscope            +listcmds          +reltime           +wildignore
+cursorbind        +localmap          +rightleft         +wildmenu
+cursorshape       -lua               -ruby              +windows
+dialog_con        +menu              +scrollbind        +writebackup
+diff              +mksession         +signs             -X11
+digraphs          +modify_fname      +smartindent       -xfontset
-dnd               +mouse             +startuptime       -xim
-ebcdic            -mouseshape        +statusline        -xpm
+emacs_tags        +mouse_dec         -sun_workshop      -xsmp
+eval              -mouse_gpm         +syntax            -xterm_clipboard
+ex_extra          -mouse_jsbterm     +tag_binary        -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L/usr/local/lib -o vim        -lm -lncurses  -liconv -framework AppKit     -L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin -lpython3.7m -ldl -framework CoreFoundation
```

<!-- Paste just the first two lines of :version here. -->

Operating System: macOS Mojave

## What went wrong

`:ALEInfo` show all kinds of errors being detected for C++ code, but does not display them anywhere. I'm using the following vimrc:

```
if &compatible
    set nocompatible
endif

set runtimepath+=$HOME/.vim/bundle/repos/github.com/Shougo/dein.vim

call dein#begin('$HOME/.vim/bundle/')
" Let dein manage dein
call dein#add('$HOME/.vim/bundle/repos/github.com/Shougo/dein.vim')

if v:version >= 800
    call dein#add('w0rp/ale', {'on_ft': ['python', 'c', 'cpp']})
    let g:ale_python_pylint_executable = 'python'
    let g:ale_set_loclist = 0
    let g:ale_set_quickfix = 1
    let g:ale_lint_on_text_changed = 'never'
    let g:ale_lint_on_insert_leave = 0
    let g:ale_lint_on_enter = 0
    let g:ale_c_parse_compile_commands = 1

    let g:ale_linters = {
                \ 'python': ['pycodestyle', 'flake8', 'pylint', 'autopep8'],
                \ 'cpp': ['clang-format', 'cppcheck'],
                \}

    let g:ale_fixers = {
                \ 'python': ['add_blank_lines_for_python_control_statements',
                \            'autopep8', 'isort', 'yapf', 'black'],
                \ 'cpp': ['clang-format', 'cppcheck']
                \}

    let g:ale_linters_explicit=1
    let g:ale_sign_column_always = 1
    let g:ale_list_window_size = 15
endif

syntax enable
```

## Reproducing the bug

<!-- Write a list of steps below. -->

1. Open Cpp file (`compile_commands.json` is working directory)
2. Run `ALELint`
3. Look at `ALEInfo`,  which shows `cppcheck` output all kinds of issues
4. Nothing in gutter, quickfix window or location list

### :ALEInfo
```

 Current Filetype: cpp
Available Linters: ['ccls', 'clang', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder', 'gcc']
   Linter Aliases:
'gcc' -> ['g++']
  Enabled Linters: ['cppcheck']
 Suggested Fixers: 
  'clang-format' - Fix C/C++ and cuda files with clang-format.
  'clangtidy' - Fix C/C++ and ObjectiveC files with clang-tidy.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_cpp_cppcheck_executable = 'cppcheck'
let g:ale_cpp_cppcheck_options = '--enable=style'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'cpp': ['clang-format', 'cppcheck'], 'python': ['add_blank_lines_for_python_control_statements', 'autopep8', 'isort', 'yapf', 'black']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 0
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 0
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'never'
let g:ale_linter_aliases = {}
let g:ale_linters = {'cpp': ['clang-format', 'cppcheck'], 'python': ['pycodestyle', 'flake8', 'pylint', 'autopep8']}
let g:ale_linters_explicit = 1
let g:ale_list_vertical = 0
let g:ale_list_window_size = 15
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 0
let g:ale_set_quickfix = 1
let g:ale_set_signs = 1
let g:ale_sign_column_always = 1
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) cppcheck
(finished - exit code 0) ['/bin/bash', '-c', 'cd ''/Users/rasmus/foobar'' && ''cppcheck'' -q --language=c++ --project=''compile_commands.json'' --enable=style ''/var/folders/_p/j1qf8m7x31g3g_c6kvnbgp840000gp/T/vNL0Itz/1/main.cpp''']

<<<OUTPUT STARTS>>>
/Users/rasmus/foobar/Class.hpp:120:3: style: 'ConfigNode::operator=' should return 'ConfigNode &'. [operatorEq]
  ConfigNode operator=(const ConfigNode& other) {
  ^
[…]
<<<OUTPUT ENDS>>>

```
(I obfuscated and truncated the error list, but there's a whole lot)

> I have used the latest version of cppcheck from git hub, and I believe they changed their error format, not sure how it was before.
> 
> This will match their format as of now (from git hub):
> file: autoload/ale/handlers/cppcheck.vim
> let l:pattern = '\v^(.+):(\d+):(\d+): ([a-z]+): (.+)$'
> instead of
> let l:pattern = '\v^[(.+):(\d+)]: (([a-z]+)) (.+)$'
> 
> add text match should be incremented by 1
> \ 'text': l:match[5],
> instead of:
> \ 'text': l:match[4],
> after this change, i could see output of cppcheck in quickfix window, and on the sign bar.

Thank you, it works for me.

My current cppcheck version is 19.0, and ale works on my Ubuntu with cppcheck 18.2.

Here is the example code:

```c
int main() {
        FILE *fp = fopen("a", "w");
        return 0;
}
```

Cppcheck 1.90:

```
Checking test.c ...
test.c:5:2: error: Resource leak: fp [resourceLeak]
 return 0;
 ^
```

Cppcheck 1.82:

```
Checking test.c ...
[test.c:5]: (error) Resource leak: fp
```

--- a/ale_linters/c/cppcheck.vim
+++ b/ale_linters/c/cppcheck.vim
@@ -10,9 +10,11 @@ function! ale_linters#c#cppcheck#GetCommand(buffer) abort
     let l:buffer_path_include = empty(l:compile_commands_option)
     \   ? ale#handlers#cppcheck#GetBufferPathIncludeOptions(a:buffer)
     \   : ''
+    let l:template = ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
 
     return l:cd_command
     \   . '%e -q --language=c'
+    \   . l:template
     \   . ale#Pad(l:compile_commands_option)
     \   . ale#Pad(ale#Var(a:buffer, 'c_cppcheck_options'))
     \   . l:buffer_path_include

--- a/ale_linters/cpp/cppcheck.vim
+++ b/ale_linters/cpp/cppcheck.vim
@@ -10,9 +10,11 @@ function! ale_linters#cpp#cppcheck#GetCommand(buffer) abort
     let l:buffer_path_include = empty(l:compile_commands_option)
     \   ? ale#handlers#cppcheck#GetBufferPathIncludeOptions(a:buffer)
     \   : ''
+    let l:template = ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
 
     return l:cd_command
     \   . '%e -q --language=c++'
+    \   . l:template
     \   . ale#Pad(l:compile_commands_option)
     \   . ale#Pad(ale#Var(a:buffer, 'cpp_cppcheck_options'))
     \   . l:buffer_path_include

--- a/autoload/ale/handlers/cppcheck.vim
+++ b/autoload/ale/handlers/cppcheck.vim
@@ -44,16 +44,21 @@ endfunction
 function! ale#handlers#cppcheck#HandleCppCheckFormat(buffer, lines) abort
     " Look for lines like the following.
     "
-    " [test.cpp:5]: (error) Array 'a[10]' accessed at index 10, which is out of bounds
-    let l:pattern = '\v^\[(.+):(\d+)\]: \(([a-z]+)\) (.+)$'
+    "test.cpp:974:6: error: Array 'n[3]' accessed at index 3, which is out of bounds. [arrayIndexOutOfBounds]\
+    "    n[3]=3;
+    "     ^
+    let l:pattern = '\v^(\f+):(\d+):(\d+): (\w+): (.*) \[(\w+)\]\'
     let l:output = []
 
     for l:match in ale#util#GetMatches(a:lines, l:pattern)
         if ale#path#IsBufferPath(a:buffer, l:match[1])
             call add(l:output, {
-            \   'lnum': str2nr(l:match[2]),
-            \   'type': l:match[3] is# 'error' ? 'E' : 'W',
-            \   'text': l:match[4],
+            \   'lnum':     str2nr(l:match[2]),
+            \   'col':      str2nr(l:match[3]),
+            \   'type':     l:match[4] is# 'error' ? 'E' : 'W',
+            \   'sub_type': l:match[4] is# 'style' ? 'style' : '',
+            \   'text':     l:match[5],
+            \   'code':     l:match[6]
             \})
         endif
     endfor

--- a/test/command_callback/test_c_cppcheck_command_callbacks.vader
+++ b/test/command_callback/test_c_cppcheck_command_callbacks.vader
@@ -1,7 +1,6 @@
 Before:
   call ale#assert#SetUpLinterTest('c', 'cppcheck')
-
-  let b:command_tail = ' -q --language=c --enable=style -I' . ale#Escape(ale#path#Simplify(g:dir)) .' %t'
+  let b:command_tail = ' -q --language=c --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'' --enable=style -I' . ale#Escape(ale#path#Simplify(g:dir)) .' %t'
 
 After:
   " Remove a test file we might open for some tests.
@@ -10,9 +9,8 @@ After:
     set buftype=nofile
   endif
 
-  call ale#assert#TearDownLinterTest()
-
   unlet! b:command_tail
+  call ale#assert#TearDownLinterTest()
 
 Execute(The executable should be configurable):
   AssertLinter 'cppcheck', ale#Escape('cppcheck') . b:command_tail
@@ -28,6 +26,7 @@ Execute(cppcheck for C should detect compile_commands.json files):
   \ ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --project=' . ale#Escape('compile_commands.json')
   \   . ' --enable=style %t'
 
@@ -38,6 +37,7 @@ Execute(cppcheck for C should detect compile_commands.json files in build direct
   \ ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/with_build_dir'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --project=' . ale#Escape(ale#path#Simplify('build/compile_commands.json'))
   \   . ' --enable=style %t'
 
@@ -47,6 +47,7 @@ Execute(cppcheck for C should include file dir if compile_commands.json file is 
   AssertLinter 'cppcheck',
   \   ale#Escape('cppcheck')
   \   . ' -q --language=c'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --enable=style'
   \   . ' -I' . ale#Escape(ale#path#Simplify(g:dir . '/cppcheck_paths'))
   \   . ' %t'
@@ -61,6 +62,7 @@ Execute(cppcheck for C should ignore compile_commands.json file if buffer is mod
   \   ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --enable=style'
   \   . ' -I' . ale#Escape(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ' %t'

--- a/test/command_callback/test_cpp_cppcheck_command_callbacks.vader
+++ b/test/command_callback/test_cpp_cppcheck_command_callbacks.vader
@@ -1,6 +1,6 @@
 Before:
   call ale#assert#SetUpLinterTest('cpp', 'cppcheck')
-  let b:command_tail = ' -q --language=c++ --enable=style -I' . ale#Escape(ale#path#Simplify(g:dir)) .' %t'
+  let b:command_tail = ' -q --language=c++ --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'' --enable=style -I' . ale#Escape(ale#path#Simplify(g:dir)) .' %t'
 
 After:
   " Remove a test file we might open for some tests.
@@ -26,6 +26,7 @@ Execute(cppcheck for C++ should detect compile_commands.json files):
   \ ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c++'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --project=' . ale#Escape('compile_commands.json')
   \   . ' --enable=style %t'
 
@@ -36,6 +37,7 @@ Execute(cppcheck for C++ should detect compile_commands.json files in build dire
   \ ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/with_build_dir'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c++'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --project=' . ale#Escape(ale#path#Simplify('build/compile_commands.json'))
   \   . ' --enable=style %t'
 
@@ -45,6 +47,7 @@ Execute(cppcheck for C++ should include file dir if compile_commands.json file i
   AssertLinter 'cppcheck',
   \   ale#Escape('cppcheck')
   \   . ' -q --language=c++'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --enable=style'
   \   . ' -I' . ale#Escape(ale#path#Simplify(g:dir . '/cppcheck_paths'))
   \   . ' %t'
@@ -59,6 +62,7 @@ Execute(cppcheck for C++ should ignore compile_commands.json file if buffer is m
   \   ale#path#CdString(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ale#Escape('cppcheck')
   \   . ' -q --language=c++'
+  \   . ' --template=''{file}:{line}:{column}: {severity}:{inconclusive:inconclusive:} {message} [{id}]\\n{code}'''
   \   . ' --enable=style'
   \   . ' -I' . ale#Escape(ale#path#Simplify(g:dir . '/cppcheck_paths/one'))
   \   . ' %t'

--- a/test/handler/test_cppcheck_handler.vader
+++ b/test/handler/test_cppcheck_handler.vader
@@ -10,19 +10,29 @@ Execute(Basic errors should be handled by cppcheck):
   AssertEqual
   \ [
   \   {
-  \     'lnum': 5,
+  \     'lnum': 974,
+  \     'col' : 6,
   \     'type': 'E',
-  \     'text': 'Array ''a[10]'' accessed at index 10, which is out of bounds',
+  \     'sub_type': '',
+  \     'text': 'Array ''n[3]'' accessed at index 3, which is out of bounds.',
+  \     'code': 'arrayIndexOutOfBounds'
   \   },
   \   {
-  \     'lnum': 7,
+  \     'lnum': 1185,
+  \     'col' : 10,
   \     'type': 'W',
-  \     'text': 'Some other problem',
+  \     'sub_type': 'style',
+  \     'text': 'The scope of the variable ''indxStr'' can be reduced.',
+  \     'code': 'variableScope'
   \   },
   \ ],
   \ ale#handlers#cppcheck#HandleCppCheckFormat(bufnr(''), [
-  \ '[test.cpp:5]: (error) Array ''a[10]'' accessed at index 10, which is out of bounds',
-  \ '[test.cpp:7]: (warning) Some other problem',
+  \ 'test.cpp:974:6: error: Array ''n[3]'' accessed at index 3, which is out of bounds. [arrayIndexOutOfBounds]\',
+  \ '    n[3]=3;',
+  \ '     ^',
+  \ 'test.cpp:1185:10: style: The scope of the variable ''indxStr'' can be reduced. [variableScope]\',
+  \ '    char indxStr[16];',
+  \ '         ^',
   \ ])
 
 Execute(Problems from other files should be ignored by cppcheck):
@@ -32,5 +42,7 @@ Execute(Problems from other files should be ignored by cppcheck):
   \ [
   \ ],
   \ ale#handlers#cppcheck#HandleCppCheckFormat(bufnr(''), [
-  \ '[bar.cpp:5]: (error) Array ''a[10]'' accessed at index 10, which is out of bounds',
+  \ 'bar.cpp:974:6: error: Array ''n[3]'' accessed at index 3, which is out of bounds. [arrayIndexOutOfBounds]\',
+  \ '    n[3]=3;',
+  \ '     ^',
   \ ])

